### DISCUSSION TITLE
Is it possible to run "prisma generate" without internet connection?

### DISCUSSION THREAD

**Discussion by KaiQiu9527** [prisma/prisma#21886] [Q&A] [answered]
### Question

Hi! I'm running prisma on a machine, which is without internet connection. 
When I run "npx prisma generate", it needs  to download the engine from internet. Even though I upload the file “libquery_engine-debian-openssl-1.1.x.so.node” and set env parameter PRISMA_QUERY_ENGINE_LIBRARY, it still download debian-openssl-1.1.x/schema-engine.sha256, which I have no idea how to set the env parameter for it.
![image](https://github.com/prisma/prisma/assets/16130943/c470478d-e6c4-464f-974f-9d0d298caeec)

Can I just generate it offline? I have check the discussion https://github.com/prisma/prisma/discussions/15865 but it doesn't solve my problem.

Thank you!

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by mimo-lbp** [comment, marked as answer]
Hello i have the same problem as you, i'm in a company under proxy and i wanted to setup a Gitlab CI/CD with my Nextjs Project. As i can't download the binary at the url address " https://binaries.prisma.sh/all_commits/aebc046ce8b88ebbcb45efe31cbe7d06fd6abc0a/rhel-openssl-1.1.x/schema-engine.gz.sha256" because the pod can't access to internet,  i would donwload mannually the query engine to put it in my folder. Like this : 
![image](https://github.com/prisma/prisma/assets/136328516/6cc1626d-0c7b-421e-9184-b224e19bbe7b) 
and after my .env : 
![image](https://github.com/prisma/prisma/assets/136328516/558f5a56-377f-484e-99d5-d2edb074afd6)

Then the problem is in the gitlab pipeline, Prisma always wants to download via the url the binaries. I read many times the docs but i cannot setup properly this. So i ask help please.
![image](https://github.com/prisma/prisma/assets/136328516/3018b90d-1271-4ff8-9ee7-4c7f9ff43d2e)
![image](https://github.com/prisma/prisma/assets/136328516/495de720-63bf-4fcc-855b-2cb6e31abd21)

Hope someone can help us  🙏

**Reply by janpio** [reply]
This is not trying to download the engine file, but a checksum (see the `.sha256` at the end of the file). This is because of a bug: https://github.com/prisma/prisma/issues/12593

**Reply by mimo-lbp** [reply]
So, im stuck with this problem for the moment ?

**Reply by janpio** [reply]
I am not aware of a simple workaround, unfortunately.
If you manage to track down where the problem comes from, PRs are welcome.

**Reply by mimo-lbp** [reply]
but i dont understand one thing, in my error above, Prisma wants to download the schema-engine so i put it in my project like this
![image](https://github.com/prisma/prisma/assets/136328516/91ac4219-3c8e-4634-b884-9d513ad1e193)
my new env : 
![image](https://github.com/prisma/prisma/assets/136328516/e4560d26-c919-4085-b6e9-23c7a49ebdfb)

and now the prisma:generate is ok but there is an another error in my npm run build  
![image](https://github.com/prisma/prisma/assets/136328516/c308a8ff-9e0e-43e6-9847-a0e59a2b95c4)

the thing i want to understand is : today prisma can't work with local query-engine and schema-engine ?

**Reply by janpio** [reply]
Yes it can. 

The error message here indicates that it can not find this file under the path you provided. Try an absolut path for example if that works. Maybe the file is loaded relative to another base than expected here.

**Reply by KaiQiu9527** [reply]
Hi @mimo-lbp ! Thanks for your info!
I read some of the issues and found that how to avoid downloading the checksum file: set the env param
`ENV PRISMA_ENGINES_CHECKSUM_IGNORE_MISSING 1`
I set it in dockerfile, and for the linux should be 
`export PRISMA_ENGINES_CHECKSUM_IGNORE_MISSING=1`
It works for me! While it looks like depend on the verison of prisma.
FYI: https://github.com/prisma/prisma/issues/19666

**Reply by janpio** [reply]
Yep, just rememberd that I was mistaken and you can _skip_ that check: https://www.prisma.io/docs/reference/api-reference/environment-variables-reference#prisma_engines_checksum_ignore_missing

**Reply by mimo-lbp** [reply]
Ty for your responses @KaiQiu9527 and @janpio. 
I already set the PRISMA_ENGINES_CHECKSUM_IGNORE_MISSING 1 in my env. 
And in local it works (prisma generate and npm run build), so i dont understand why in local it works and not in my pipeline gitlab.
In the pipeline's error, it say "The Prisma engines do not seem to be compatible with your system." and i checked my config 

Local : node 20.9.0, npm 10.1.0
Pipeline node 16.20.2, npm 8.19.4

Maybe the node version is to low to work normally ? My next try is to set the same node/npm version 

The env that i used in local where npm run build worked
![image](https://github.com/prisma/prisma/assets/136328516/29098eec-d655-49a5-a9ac-81279dc80dfe)

**Reply by janpio** [reply]
The "not compatible" message might be a red herring, Prisma potentially only has 1 error message when the loading of the files does not work.

In the screenshot above it says "No such file or directory". So I think for some reason it can just not find your `.dll.node` file.

**Reply by mimo-lbp** [reply]
![image](https://github.com/prisma/prisma/assets/136328516/7eb97e60-d06c-419f-8884-d12cb3a1fc92)
An idea what i write as path ? /opt/app-root/src/query... ?

**Reply by mimo-lbp** [reply]
I tested with and i have a new error
![image](https://github.com/prisma/prisma/assets/136328516/82f557a7-116d-4bf5-b262-c6d8e6679e9c)

I think (im newbie in devops) i used query engine for windows and in local im using windows but in my pipeline its think its a Linux OS so its maybe because it doesnt work ?
![image](https://github.com/prisma/prisma/assets/136328516/11a3563c-f91f-4362-a618-92a29b198ea6)

**Reply by janpio** [reply]
No idea, sorry. Quick googling does not give helpful information.

Is your pipeline actually running on Windows?

**Reply by mimo-lbp** [reply]
No my pipeline is running on Linux, that's the difference

**Reply by janpio** [reply]
Then you of course have to provide a Prisma engine that was built for Linux, not the one for Windows.
You can use `binaryTargets` locally to make it download multiple engines and to get these download URLs so you can provide the files locally.

**Reply by SpinBoxx** [reply]
Hello, im the same guy that @mimo-lbp (i using personnal account). 
So, after many tries, i successfully push my prisma project on production with Gitlabcicd under enterprise proxy. To help guys were like me :

1. Locally, i downloaded query/schema engines for linux OS with the command npx prisma init by default
2. I copied the query/schema engines that were in the folder node_modules/@prisma/engines in my root folder
![image](https://github.com/prisma/prisma/assets/45215351/b28a31a2-3df4-4cb2-85ec-045eab4069eb)
![image](https://github.com/prisma/prisma/assets/45215351/52bb1354-b20d-4efe-b572-b5eb5263fe27)

4. I setted my .env like this 
![image](https://github.com/prisma/prisma/assets/45215351/37888c62-1312-4f5f-945e-b52861c4dce5)

5. I added new commands in my package.json (its usefull when u switch prod/dev): 
![image](https://github.com/prisma/prisma/assets/45215351/6da08f4e-0f09-4ffe-a4f4-502a05182d9f)

6. I using only the command prisma:gen (npx prisma generate) in my Giltab-ci/cd to generate Client models
